test/aleph/http_test.clj

```diff
@@ -1464,6 +1464,162 @@
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http1])))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "Some desired HTTP versions are not part of ALPN config." (ex-message result)))))
+
+      (testing "with different preference order in ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http1 :http2])))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "Desired HTTP version preference order differs from ALPN config." (ex-message result)))))
+
+      (testing "with extra HTTP versions in the ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http1]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http1 :http2])))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "ALPN config contains more HTTP versions than desired." (ex-message result)))))
+
+      (testing "with matching ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http2 :http1])))}})]
+          (is (= :success result))))
+
+      (testing "with no ALPN config but desiring HTTP/2"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context test-ssl/client-ssl-context}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "No ALPN supplied, but requested non-HTTP/1 versions that require ALPN." (ex-message result)))))
```

Comment:
We should probably specify in the error message that they requested _secure_ HTTP2. 

On the rare chance they're using h2c, there will be no ALPN (or TLS, for that matter).

Reply:
Good idea. While at it, I also made it explicit that this is about the supplied `SslContext`.